src/native/mqtt5_listener.c

```diff
@@ -0,0 +1,599 @@
+    }
+
+    if (!publish) {
+        AWS_LOGF_ERROR(AWS_LS_MQTT5_GENERAL, "publishReceived function: invalid publish packet");
+        return false;
+    }
+
+    /********** JNI ENV ACQUIRE **********/
+    JavaVM *jvm = java_listener->jvm;
+    JNIEnv *env = aws_jni_acquire_thread_env(jvm);
+    if (env == NULL) {
+        /* If we can't get an environment, then the JVM is probably shutting down.  Don't crash. */
+        AWS_LOGF_ERROR(AWS_LS_MQTT5_GENERAL, "publishReceived function: could not get env");
+        return false;
+    }
+
+    /* Calculate the number of references needed */
+    size_t references_needed = 0;
+    {
+        /* One reference is needed for the PublishReturn */
+        references_needed += 1;
+
+        /* A Publish packet will need 5 references at minimum */
+        references_needed += 5;
+        /* Optionals */
+        s_aws_count_allocation(publish->content_type, &references_needed);
+        s_aws_count_allocation(publish->correlation_data, &references_needed);
+        s_aws_count_allocation(publish->message_expiry_interval_seconds, &references_needed);
+        s_aws_count_allocation(publish->response_topic, &references_needed);
+        s_aws_count_allocation(publish->topic_alias, &references_needed);
+        s_aws_count_allocation(publish->payload_format, &references_needed);
+        /* Add user properties and subscription identifiers */
+        references_needed += publish->user_property_count * 2;
+        references_needed += 1; /* Add 1 for array to hold user properties */
+        if (publish->subscription_identifier_count > 0) {
+            references_needed += publish->subscription_identifier_count;
+            references_needed += 1; /* Add 1 for array */
+        }
+    }
+
+    /**
+     * Push a new local frame so any local references we make are tied to it. Then we can pop it to free memory.
+     */
+    jint local_frame_result = (*env)->PushLocalFrame(env, (jint)references_needed);
+    if (local_frame_result != 0) {
+        s_aws_mqtt5_listener_log_and_throw_exception(
+            env,
+            "publishReceived function: could not push local JNI frame with 12 allocation minimum!",
+            AWS_ERROR_INVALID_STATE);
+        goto clean_up;
+    }
+
+    /* The return result */
+    jobject publish_packet_return_data;
+
+    /* Make the PublishPacket */
+    jobject publish_packet_data = s_aws_mqtt5_client_create_jni_publish_packet_from_native(env, publish);
+    if (publish_packet_data == NULL) {
+        goto clean_up;
+    }
+
+    /* Make the PublishReturn struct that will hold all of the data that is passed to Java */
+    publish_packet_return_data = (*env)->NewObject(
+        env,
+        mqtt5_publish_return_properties.return_class,
+        mqtt5_publish_return_properties.return_constructor_id,
+        publish_packet_data);
+    aws_jni_check_and_clear_exception(env); /* To hide JNI warning */
+
+    if (java_listener->jni_listener_publish_events) {
```

Comment:
Extremely minor nitpick: I'd move this check earlier in the function and escape early if the function does not exist. That way there's not a need to do additional allocations and calculations when not necessary.